src/main/java/dev/morling/onebrc/CalculateAverage_fwbrasil.java

```diff
@@ -0,0 +1,383 @@
+                hash += b;
+                hash += (hash << 10);
+                hash ^= (hash >> 6);
+            }
+            hash += (hash << 3);
+            hash ^= (hash >> 11);
+            hash += (hash << 15);
+            var r = Math.abs(hash * PRIME);
+            return r;
+        }
+
+
+        int readValue() {
+            var value = Integer.MAX_VALUE;
+            var negative = false;
+            if (read(pos) == '-') {
+                negative = true;
+                pos++;
+            }
+            byte maybeDot = read(pos + 1);
+            if (maybeDot == '.') {
+                value = read(pos) * 10 + read(pos + 2) - 528;
+                pos += 3;
+            }
+            else {
+                value = read(pos) * 100 + maybeDot * 10 + read(pos + 3) - 5328;
+                pos += 4;
+            }
+            return negative ? -value : value;
+        }
+    }
+
+    final class Results {
+
+        static final long TABLE_BASE_OFFSET = UNSAFE.arrayBaseOffset(long[].class);
+        static final long TABLE_SCALE = UNSAFE.arrayIndexScale(long[].class);
+        static final int KEYS_SIZE = (int) Math.pow(2, 15);
+        static final int KEYS_MASK = KEYS_SIZE * 2 - 1;
+        static final int ENTRIES_SIZE = 10_000;
+
+        // keyHash | stateIdx
+        final long[] keys = new long[KEYS_SIZE * 2];
+
+        // keyStart | keyEnd
+        final long[] meta = new long[ENTRIES_SIZE * 2];
+
+        // sum | packed(min | max | count)
+        final long[] values = new long[ENTRIES_SIZE * 2];
+
+        final Packed packed = new Packed();
+
+        int nextEntry = 0;
+
+        public void add(long keyStart, long keyEnd, long hash, int value) {
+            var idx = findKey(keyStart, keyEnd, hash);
+            update(value, idx);
+        }
+
+        private void update(int value, int idx) {
+            add(values, idx, value);
+            long packed2 = get(values, idx + 1);
+            packed.decode(packed2);
+            packed.update(value);
+            set(values, idx + 1, packed.encode());
+        }
+
+        public void merge(Results other) {
+            for (int i = 0; i < keys.length; i += 2) {
+                var hash = 0L;
+                if ((hash = get(other.keys, i)) != 0) {
+                    var idx = (int) get(other.keys, i + 1);
+                    var keyStart = get(other.meta, idx);
+                    var keyEnd = get(other.meta, idx + 1);
+                    var sum = get(other.values, idx);
+                    other.packed.decode(get(other.values, idx + 1));
+
+                    idx = findKey(keyStart, keyEnd, hash);
+                    add(values, idx, sum);
+                    packed.decode(get(values, idx + 1));
+                    packed.merge(other.packed);
+                    set(values, idx + 1, packed.encode());
+                }
+            }
+        }
+
+        int findKey(long keyStart, long keyEnd, long hash) {
```

Comment:
Where do you deal with hash collisions?

Reply:
The goal was to find a good hash function without conflicts in the long range for the workload and tests. Is that not allowed? I'm getting access to a linux box today to work on the benchmark in case you're still seeing differences.

Reply:
As per the rules, it must work for any station names of the allowed length:

> Q: Can I make assumptions on the names of the weather stations showing up in the data set?
A: No, while only a fixed set of station names is used by the data set generator, any solution should work with arbitrary UTF-8 station names (for the sake of simplicity, names are guaranteed to contain no ; or \n characters).

We can't test for all potential hash collisions, so the test set is just the minimum bar to pass, but it must be ensured that an implementation works for any valid data set.

Reply:
I'll change the implementation but it'd be nice to have more clear guidelines if you're planning to have another round next year. Most (if not all) of the top contenders have major adaptations to the workload:

1. **Fixed hash table size**: People tune the size to make the workload and tests pass but an execution with more keys could make them enter an infinite loop.
2. **Numeric overflows**: Most implementations use more compact types for things like the sum. It's ok for the current dataset but it could overflow in larger workloads.
3. **Format assumptions**: Several contenders assume the format of the specific workload and would fail if a temperature >= 100 was in the dataset for example.

I'd say finding a good hash function that works well for city names in general in the `long` range seems more reasonable if compared to those assumptions. If latency and performance was a real constraint, I believe that could be a viable tradeoff in a production workload.

Reply:
I hear you in regards to more precisely defined constraints, and yes, it's certainly one of my personal learnings for the next time (no promises yet ;). That said, most of the things you describe in fact are described already:

> Fixed hash table size

There can be 10K stations at most, so a fixed size is fine. You must be able to handle any valid names though, not only those of the example data set.

> Numeric overflows

Do you have an example for that? I suppose a 1B data set with measurements only for one station may overflow an int. Indeed there's no test for that.

> Format assumptions

Temperatures are defined to be between -99.9 and 99.9 (inclusive).